### PR TITLE
Dynamic Matchmaking

### PR DESCRIPTION
**Pull request by kev306**
Updated matchmaking to allow for 6-8p matches
Waits 10seconds before finding a match from when a user last joins
Prioritises larger lobbies

Addresses issue #613

### CODE REVIEW

File: src/sockets/sockets.ts [hunk 1]
```diff
@@ -2278,7 +2279,34 @@ function matchFound(usernames: string[]): void {
...
+          break;
+        case 8:
+          room.startGame([
+            Role.Merlin,
+            Role.Percival,
+            Role.Assassin,
+            Role.Morgana,
+            Card.LadyOfTheLake,
+          ]);
+          break;
```
Comment (vck3000) [inline src/sockets/sockets.ts:2308]: Add a default clause and throw here in case we get something unexpected.

File: src/sockets/tests/matchmakingQueue.test.ts [hunk 2]
```diff
@@ -24,9 +24,11 @@ describe('MatchmakingQueue', () => {
...
 
     matchmakingQueue.addUser(getDefaultQueueEntry('6'));
-    const expectUsernames = ['1', '2', '3', '4', '5', '6'];
-    expect(matchFoundCallback).toHaveBeenCalledWith(expectUsernames);
-    expect(matchmakingQueue.getNumInQueue()).toEqual(0);
+    setTimeout(() => {
+      const expectUsernames = ['1', '2', '3', '4', '5', '6'];
+      expect(matchFoundCallback).toHaveBeenCalledWith(expectUsernames);
+      expect(matchmakingQueue.getNumInQueue()).toEqual(0);
+    }, 10000);
```
Comment (vck3000) [inline src/sockets/tests/matchmakingQueue.test.ts:31]: Let’s not add time delays into the test code itself.

There is a way for you to fast forward time in the test code. Look it up in jest :). 

https://jestjs.io/docs/timer-mocks

I’ve used it in a test file. You should be able to find examples.

File: src/views/about.ejs [hunk 3]
```diff
@@ -64,7 +64,7 @@
   <br />
   <img
     id="myAvatar"
-    src="https://cdn.discordapp.com/attachments/430166478193688597/481009331622510602/pronub-res.png"
+    src="https://i.ibb.co/jz94y0d/Pro-Nub-res.png"
```
Comment (vck3000) [inline src/views/about.ejs:67]: Split this into a separate PR pls.

File: src/sockets/matchmakingQueue.ts [hunk 4]
```diff
@@ -14,20 +18,33 @@ export class QueueEntry {
...
     this.matchFoundCallback = matchFoundCallback;
   }
 
+  private startQueueCheckTimer() {
+    if (this.queueCheckTimer !== null) {
+      clearTimeout(this.queueCheckTimer);
+    }
+    this.queueCheckTimer = setTimeout(() => {
+      this.checkQueue();
+    }, TEN_SECONDS);
```
Comment (vck3000) [inline src/sockets/matchmakingQueue.ts:35]: Let’s call this variable “MM_JOIN_WINDOW”

File: src/sockets/matchmakingQueue.ts [hunk 5]
```diff
@@ -14,20 +18,33 @@ export class QueueEntry {
...
     if (this.getQueueUsernames().includes(queueEntry.username)) {
       return false;
     }
 
     this.queue.push(queueEntry);
-    this.checkQueue();
+
+    if (this.queue.length >= MAX_PLAYERS_PER_GAME) {
+      this.checkQueue();
+    } else this.startQueueCheckTimer();
```
Comment (vck3000) [inline src/sockets/matchmakingQueue.ts:47]: Surround this guy in braces pls.

File: src/sockets/matchmakingQueue.ts [hunk 6]
```diff
@@ -65,46 +82,47 @@ export class MatchmakingQueue {
...
-    const combinations = MatchmakingQueue.getCombinations(
-      this.queue,
-      NUM_PLAYERS_PER_GAME,
-    );
-
-    for (const combination of combinations) {
-      const combinationUsernames = combination.map((entry) => entry.username);
-      let combinationFailed = false;
+    for (let i = MAX_PLAYERS_PER_GAME; i >= MIN_PLAYERS_PER_GAME; i--) {
+      const combinations = MatchmakingQueue.getCombinations(this.queue, i);
```
[the remark this answers is not in the document]
Comment (kev306) [reply src/sockets/matchmakingQueue.ts]: Done in https://github.com/vck3000/ProAvalon/pull/627/commits/de91bdc1817085b057879af44f1d1cd0da1b4261

File: src/sockets/matchmakingQueue.ts [hunk 7]
```diff
@@ -65,16 +84,25 @@ export class MatchmakingQueue {
   }
 
   private checkQueue(): void {
-    if (this.queue.length < NUM_PLAYERS_PER_GAME) {
+    if (this.queue.length < MIN_PLAYERS_PER_GAME) {
       return;
     }
 
+    for (
```
Comment (vck3000) [inline src/sockets/matchmakingQueue.ts:91]: Can you add a comment saying why we want to go in reverse order. To prioritise larger games.

File: src/sockets/matchmakingQueue.ts [hunk 8]
```diff
@@ -65,16 +84,25 @@ export class MatchmakingQueue {
...
+      let gameSize = MAX_PLAYERS_PER_GAME;
+      gameSize >= MIN_PLAYERS_PER_GAME;
+      gameSize--
+    ) {
+      if (this.checkValidCombinations(gameSize)) {
```
Comment (vck3000) [inline src/sockets/matchmakingQueue.ts:96]: checkValidGameSize maybe?

Comment (vck3000) [reply src/sockets/matchmakingQueue.ts:96]: Thing is the function is also calling back to sockets so doesn’t feel suitable to call it a check. 

How about TryMatchGameSize

File: src/sockets/matchmakingQueue.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (vck3000) [reply src/sockets/matchmakingQueue.ts]: MatchmakingQueue.getCombinations() will return []
